api/utils/tlsutils/tlsutils.go

```diff
@@ -36,3 +40,67 @@ func ParseCertificatePEM(bytes []byte) (*x509.Certificate, error) {
 	}
 	return cert, nil
 }
+
+// ContextDialer represents network dialer interface that uses context
+type ContextDialer interface {
+	// DialContext is a function that dials the specified address
+	DialContext(ctx context.Context, network, addr string) (net.Conn, error)
+}
+
+// TLSDial dials and establishes TLS connection using custom dialer
+// is similar to tls.DialWithDialer
+// Note: function taken from lib/utils/tlsdial.go
+func TLSDial(ctx context.Context, dialer ContextDialer, network, addr string, tlsConfig *tls.Config) (*tls.Conn, error) {
+	if tlsConfig == nil {
+		return nil, trace.BadParameter("tls config must be specified")
+	}
+
+	plainConn, err := dialer.DialContext(ctx, network, addr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	colonPos := strings.LastIndex(addr, ":")
+	if colonPos == -1 {
+		colonPos = len(addr)
+	}
+	hostname := addr[:colonPos]
+
+	// If no ServerName is set, infer the ServerName
+	// from the hostname we're connecting to.
+	if tlsConfig.ServerName == "" {
+		// Make a copy to avoid polluting argument or default.
+		c := tlsConfig.Clone()
+		c.ServerName = hostname
+		tlsConfig = c
+	}
+
+	conn := tls.Client(plainConn, tlsConfig)
+	errC := make(chan error, 1)
+	go func() {
+		err := conn.HandshakeContext(ctx)
+		errC <- err
+	}()
+
+	select {
+	case err := <-errC:
+		if err != nil {
+			plainConn.Close()
+			return nil, trace.Wrap(err)
+		}
+	case <-ctx.Done():
+		plainConn.Close()
+		return nil, trace.BadParameter("tls handshake has been canceled due to timeout")
+	}
```

Comment:
Do we need this? Won't `conn.HandshakeContext` fail by itself if context times out?

Reply:
Hm,  I just copied it over from `tlsdial.go`, but yep, seems to be redundant 🤔 Removing.